src/AutoRest.CSharp/Mgmt/Decorator/InheritanceChoser.cs

```diff
@@ -27,54 +27,99 @@ private static IList<System.Type> GetReferenceClassCollection()
             return assembly.GetTypes().Where(t => t.GetCustomAttributes(false).Where(a => a.GetType() == typeof(ReferenceTypeAttribute)).Count() > 0).ToList();
         }
 
-        public static void RebuildModelInheritance(Dictionary<Schema, TypeProvider> schemaMap, Dictionary<Schema, TypeProvider> resourceSchemaMap)
+        private static List<System.Type> GetOrderedList()
         {
-            var typeOverrideMap = new Dictionary<CSharpType, CSharpType>();
+            var referenceClasses = ConvertGenericType(GetReferenceClassCollection());
+            var trees = GetTrees(referenceClasses);
+            var output = new List<System.Type>();
+            foreach (var root in trees)
+            {
+                var treeNodes = new List<System.Type>();
+                Queue<Node> queue = new Queue<Node>();
+                queue.Enqueue(root);
+                while (queue.Count != 0)
+                {
+                    Node tempNode = queue.Dequeue();
+                    treeNodes.Add(tempNode.type);
+                    List<Node> tempChilren = tempNode.children;
+                    if (tempChilren != null)
+                    {
+                        int childNum = tempChilren.Count;
+                        while (childNum > 0)
+                        {
+                            queue.Enqueue(tempChilren[childNum - 1]);
+                            childNum--;
+                        }
+                    }
+                }
+                treeNodes.Reverse();
+                output.AddRange(treeNodes);
+            }
+            return output;
+        }
 
-            RebuildExactModelInheritance(schemaMap);
-            RebuildExactModelInheritance(resourceSchemaMap);
+        private static IList<System.Type> ConvertGenericType(IList<System.Type> referenceClassCollection)
+        {
+            for (int i = 0; i < referenceClassCollection.Count; i++)
+            {
+                referenceClassCollection[i] = referenceClassCollection[i].MakeGenericType(referenceClassCollection[i].GetResourceIdentifierType());
```

Comment:
this should be coming from the customeAttribute property we added in Azure.ResourceManager.Core.ReferenceTypeAttribute